foundationdb/assets/configuration/spec.yaml

```diff
@@ -0,0 +1,50 @@
+name: FoundationDB
+files:
+- name: foundationdb.yaml
+  options:
+  - template: init_config
+    options:
+    - template: init_config/default
+  - template: instances
+    options:
+    - template: instances/default
+    - name: base_command
```

Comment:
This feels like a bad approach because it requires the agent to run locally on the same machine as foundation db plus having to execute a subrpcess and parse the result is usually error prone so it should only be done when there is no other alternative. Typically integrations use a library or API to communicate with services and in the case of foundation db there seems to be an officially apple supported python library https://apple.github.io/foundationdb/api-python.html.

Reply:
Strongly agree, here is how: https://apple.github.io/foundationdb/mr-status.html

Reply:
> This feels like a bad approach because it requires the agent to run locally on the same machine as foundation db

This is incorrect; it requires the FoundationDB *client binaries* to be installed on the machine that the agent runs on. However, the Python library also requires these very same client binaries. It seems the library and CLI are distributed together, so if you have the binaries that the Python library wants installed, you have the CLI too. Thus using the Python library means the integration would have more dependencies, not less. The link to the Python library you gave also notes that the additional issue of having to make sure `libffi` is installed too depending on Python version:

> The FoundationDB Python API is compatible with Python 2.7 - 3.7. You will need to have a Python version within this range on your system before the FoundationDB Python API can be installed. Also please note that Python 3.7 no longer bundles a full copy of libffi, which is used for building the _ctypes module on non-macOS UNIX platforms. Hence, if you are using Python 3.7, you should make sure libffi is already installed on your system.

Also, about this:

> having to execute a subrpcess and parse the result is usually error prone

While I'd certainly agree in the case of parsing text output that could change at any time, here the output is JSON - in fact, the very same we'd get if we used the Python library and queried the special key.

We can change this if you really insist, but it's the way it is not because we were unaware of the Python binding option, but because there seems to be no advantages to be had in return for additional moving parts.

Reply:
FYI the Agent is guaranteed to have `libffi`. Also, is Python 3.7 a hard upper limit or just what they've tested?

Reply:
> FYI the Agent is guaranteed to have libffi

OK, that's at least something if we have to go the Python way.

> Also, is Python 3.7 a hard upper limit or just what they've tested?

I've not found confirmation either way, although that line in the documentation was last edited in March 2019, so it's most likely out-dated.

Reply:
Adding coding dependencies is not an issue. Adding dependencies that the customer has to manually install is not a great experience for the user so it should be avoided. Executing subprocesses is a source of problems and should be avoided when possible

Reply:
> Adding coding dependencies is not an issue. Adding dependencies that the customer has to manually install is not a great experience for the user so it should be avoided

What I'm trying (and apparently failing) to explain is that it's not a choice of code (I assume we mean Python) dependencies *or* the FoundationDB client binaries. If one creates a clean environment (I used an Ubuntu container), does `pip install foundationdb`, and then tries to connect to a cluster, the result is:

```
Traceback (most recent call last):
  File "x.py", line 2, in <module>
    fdb.api_version(630)
  File "/usr/local/lib/python3.8/dist-packages/fdb/__init__.py", line 68, in api_version
    import fdb.impl
  File "/usr/local/lib/python3.8/dist-packages/fdb/impl.py", line 1300, in <module>
    raise Exception("Unable to locate the FoundationDB API shared library!")
Exception: Unable to locate the FoundationDB API shared library!
```

That is to say, the FoundationDB client binaries also have to be installed **regardless**. I don't believe there's a way to avoid that manual step. Using the Python library does not eliminate it, as demonstrated.

> Executing subprocesses is a source of problems and should be avoided when possible

I believe the concrete points so far have been:

* Parsing process output is fragile. This is true if the output is not meant to be machine readable. But we get JSON back; the command we're invoking was designed for obtaining machine readable status information.
* Making the user install the FoundationDB client is not ideal. This is true, but the client needs to be installed anyway because the Python binding for FoundationDB relies on it.

I'm getting the impression that no matter what arguments I put forward, there's no way this integration will get merged unless we reimplement it using the FoundationDB Python library to do the status query. If that's the case, please say so, and we'll do it; it just seems to me like a net increase in complexity (more dependencies) and things that might go wrong (such as a future incompatibility between the FoundationDB Python binding bundled in the integration and the installed client libraries).

Reply:
Can queries be made by via the CLI? For our other database integrations we allow users to define custom queries to monitor.

Reply:
We've decided to begin shipping the `foundationdb` Python library with the Agent. As far as [API versioning](https://apple.github.io/foundationdb/api-python.html#api-versioning) goes, let's add an option that defaults to the current `630`.

I'll update here when I open the dependency PR.

Reply:
Thanks for your patience here! We've added a way ship dependencies for this use case: https://github.com/DataDog/integrations-core/pull/10040

1. Nope, no need. If you ever wish to change the version, modify these files: https://github.com/DataDog/integrations-core/pull/10050
2. Yes, set [this](https://github.com/edumentab/integrations-extras/blob/bf5c223ae671fcac6b284feaf9597c04bb1cb63b/foundationdb/setup.py#L28) to `21.1.0`.
3. For regular tests you'll need to add the dependency to the `requirements-dev.txt` like [this](https://github.com/DataDog/integrations-core/blob/master/ibm_db2/requirements-dev.txt) and I suppose you'll need to create a [script](https://datadoghq.dev/integrations-core/meta/ci/#scripts) in this repo that installs the client ([example](https://github.com/DataDog/integrations-core/blob/master/.azure-pipelines/scripts/ibm_mq/linux/55_install_client.sh)). For tests [marked as E2E](https://datadoghq.dev/integrations-core/ddev/plugins/#agent-check-runner) you'll need to install the client inside of the Agent container using [environment metadata](https://datadoghq.dev/integrations-core/ddev/plugins/#environment-manager) like [this](https://github.com/DataDog/integrations-core/blob/e08d2911e65f0bd3d8b9a7e0566aadae534c94c2/ibm_mq/tests/conftest.py#L216) defined [here](https://github.com/DataDog/integrations-core/blob/e08d2911e65f0bd3d8b9a7e0566aadae534c94c2/ibm_mq/tests/common.py#L140-L144).